### PR TITLE
Make building Components & Mvc solutions work out of the box

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes https://github.com/aspnet/AspNetCore/issues/8413

### CODE REVIEW

File: src/Components/Browser.JS/Microsoft.AspNetCore.Components.Browser.JS.npmproj [hunk 1]
```diff
@@ -12,10 +12,18 @@
...
+    if the node_modules directory is missing.
+  -->
+  <Target Name="FastRestore" DependsOnTargets="Restore" Condition="!Exists('node_modules')" />
+
   <!-- Override the default 'Build' target from eng/targets/Npm.Common.targets. -->
   <Target Name="Build"
+          DependsOnTargets="FastRestore"
           Inputs="@(WebpackInputs)"
-          Outputs="dist\components.webassembly.js;dist\components.server.js">
+          Outputs="src\dist\components.webassembly.js;src\dist\components.server.js">
```
Comment (rynowak) [inline src/Components/Browser.JS/Microsoft.AspNetCore.Components.Browser.JS.npmproj]: Consider moving this to an item group

File: src/Components/Browser.JS/Microsoft.AspNetCore.Components.Browser.JS.npmproj [hunk 2]
```diff
@@ -12,10 +12,18 @@
 
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), Directory.Build.targets))\Directory.Build.targets" />
 
+  <!--
+    Invoking build[.cmd|.sh] does not restore npm projects by default. This is a performance optimization.
+    However, we do no want to ensure yarn's restored before a P2P build happens. We'll conservatively yarn install
```
Comment (pranavkm) [inline src/Components/Browser.JS/Microsoft.AspNetCore.Components.Browser.JS.npmproj]: we do ~~not~~ want to

**Review by rynowak — APPROVED** [review]

File: src/Components/Browser.JS/Microsoft.AspNetCore.Components.Browser.JS.npmproj [hunk 3]
```diff
@@ -8,14 +8,24 @@
 
   <ItemGroup>
     <WebpackInputs Include="**\*.ts" Exclude="node_modules\**" />
+    <WebPackOutputs Include="src\dist\components.webassembly.js" />
+    <WebPackOutputs Include="src\dist\components.server.js" />
```
Comment (SteveSandersonMS) [inline src/Components/Browser.JS/Microsoft.AspNetCore.Components.Browser.JS.npmproj]: Nit: for consistency, `WebpackOutputs `

**Comment by SteveSandersonMS** [comment]
I'd love to understand a bit more about this idea that it's legitimate to build Components.Server with `BuildNodeJS==false`. As far as I can tell, that just means we're making a broken build, since Components.Server can't function correctly without the compiled .js files.

If we do need to have this concept of `BuildNodeJS==false`, shouldn't we be excluding things that depend on Node-build resources (such as Components.Server) completely, instead of building a malformed and unfunctional version of them?

**Comment by SteveSandersonMS** [comment]
To be clear, I'm not questing the correctness of your changes here, @pranavkm! This looks totally right within the scope of what we already do. I'm more wondering if, since we're changing this anyway, we have the opportunity to change it more completely so it doesn't do this weird thing any more.

File: src/Components/Browser.JS/Microsoft.AspNetCore.Components.Browser.JS.npmproj (hunk 3 above)
Comment (pranavkm) [reply src/Components/Browser.JS/Microsoft.AspNetCore.Components.Browser.JS.npmproj]: I'll do it in a followup. I'd rather not wait another day to fix up casing 😄

**Comment by pranavkm** [comment]
> I'd love to understand a bit more about this idea that it's legitimate to build Components.Server with BuildNodeJS==false. As far as I can tell, that just means we're making a broken build, since Components.Server can't function correctly without the compiled .js files.

The flag is set for the linux-musl build. I think the image is missing nodejs and yarn. It's not an immediate problem since builds for the Components.Server package come out from our Windows build, and the musl build runs does not run any tests.

**Pull request merged**